Odin v3 beta

adammacias opened the pull request:
Galera, o peguei as últimas semanas para dar atenção especial para o Odin. Levei em consideração os tópicos https://github.com/wpbrasil/odin/issues/328 e https://github.com/wpbrasil/odin/issues/411 e agradeço demais o pessoal que sugeriu as mudanças e entrou na discussão. 

Segue alguns destaques dessa versão:

- Adicionado [Bootstrap 4](http://getbootstrap.com.br/) via npm.
- Adicionado [Font Awesome 5](https://fontawesome.com/) via npm
- Refatorado tarefas de automatização Grunt para Gulp.
- Adicionado comando `npm run dev` inicia modo de desenvolvimento com atualização em tempo real com [browsersync](https://browsersync.io/).
- Adicionado comando `npm run build` compila e minifica arquivos assets (CSS, JS, Imagens e Fonts) para produção `dist`.
- Adicionado comandos `npm run build:{css,js,img,fonts}` compila individualmente arquivos assets para produção `dist`.
- Adicionado  comando `npm run lang:checktextdomain` verifica text domain nos arquivos PHP do tema. 
- Adicionado  comando `npm run lang:addtextdomain` adiciona text domain nos arquivos do tema. 
- Adicionado  comando `npm run lang:makepot` cria arquivo [.POT](https://poedit.net) para internacionalização de tema. 
- Refatorado o "core" do Odin para [odin-toolkit](https://github.com/wpbrasil/odin-toolkit) plugin.
- Adicionado [TGM Plugin Activation](http://tgmpluginactivation.com/configuration/) com sugestão para instalação do [odin-toolkit](https://github.com/wpbrasil/odin-toolkit). (Merge [v3-dev](https://github.com/wpbrasil/odin/tree/v3-dev))
- Removido funções desnecessárias como a de paginação, cujo WordPress já criou-se nativamente.
- Refatorado estrutura de arquivos e diretórios em geral, como `functions/`, `assets/`, `inc/`, 'components/' e etc.

Acho que é isso pessoal, aguardando os testes e o feedback de vcs antes do lançamento! :)

![odin-3-screen1](https://user-images.githubusercontent.com/1139893/38696086-81b002ee-3e64-11e8-8504-54851550a835.gif)

## Componentes

A grande mudança do Odin3 e talvez a diferença entre os outros temas base é a ideia é "componetizar". 
 Acredito que seja uma boa ideia seguir (ou tentar) o que os frameworks JS estão fazendo :)

Exemplo do [componente/post/](https://github.com/wpbrasil/odin/tree/v3-beta/components/post). No mesmo diretório é possível encontrar o template (1), funções (2), os estilos (3) e scripts (4) do componente. Isto facilita na hora de edita-lo ou reaproveita-lo mais tarde.

1. **Template**. seria nossa "view" [`components/post/post.php`](https://github.com/wpbrasil/odin/tree/v3-beta/components/post/post.php).
2. **Funções** PHP exclusivas do compoentes em questão, veja o exemplo [`components/post/post.php`](https://github.com/wpbrasil/odin/tree/v3-beta/components/post/post.functions.php).
3. **Estilos** (SCSS) exclusivos do componente. [`components/post/post.php`](https://github.com/wpbrasil/odin/tree/v3-beta/components/post/_post.scss). Caso ainda não tenha o prefix underscore, exemplo `posts.scss` é gerado um arquivo `post.css` em `dist`, isso se for interessante link-lo na página separadamente.
4. **Scripts**. JavaScript utilizado do componente, arquivos .js é automaticamente cuspido em `dist` durante a compilação, daí é só utilizar `wp_enqueue_script()` no nosso passo (2) para add na página.

A base é esta, se deletar o diretório `components/post` o componente morre, e para reaproveita-lo em outro tema é só copiar e colar o diretório.

## Estrutura

### `/dist/`

Este diretório só é construído a partir do `npm run (dev|build)` como de costume já em outros conhecidos frameworks, detalhes dos comandos em [README.md](https://github.com/wpbrasil/odin/tree/v3-beta#getting-start).

Caso a constante [SCRIPT_DEBUG](https://codex.wordpress.org/Debugging_in_WordPress) estiver ativa, o Odin é esperto o suficiente para chamar os arquivos minificados ou de desenvolvimento.

```
/dist
├── css
|  ├── admin.css            => Estilo do Admin do WP
|  ├── admin.min.css        => Estilo do Admin do WP (produção)
|  ├── editor.css           => Estilo do editor nativo do WP
|  ├── editor.min.css       => Estilo do editor nativo do WP (produção)
|  ├── theme.css            => Estilo do Tema
|  └── theme.min.css        => Estilo do Tema (produção)
├── fonts
|  ├── fa-regular-400.eot   => Arquivo de tipografia (FontAwesome e etc)
|  ├── fa-regular-400.svg   => ...
|  ├── fa-regular-400.ttf   => ...
|  └── ...
├── img
|  ├── admin-logo-login.png => Logo do admin (Login)
|  ├── theme-icon.png       => Favicon do Tema
|  └── theme-logo.png       => Logo padrão do Tema (caso não use o personalizar do WP)
└── js
├── admin.js                => Scripts do Admin do WP
├── admin.min.js            => Scripts do Admin do WP (produção)
├── theme.js                => Scripts do Tema (produção)
└── theme.min.js            => Scripts do Tema (produção)
```

### `/assets/`

No arquivo `js/theme.js` existe a possibilidade de incluir outros arquivos nele mesmo, veja o [exemplo de como o Bootstrap](https://github.com/wpbrasil/odin/blob/v3-beta/assets/js/theme.js#L19-L38) é incluído.

O arquivo `css/theme.scss` inclui as variaveis, functions e mixins e estilos do tema e terceiros. Note que tambem inclui qualquer arquivo sass do diretório[ `/components/`](https://github.com/wpbrasil/odin/blob/v3-beta/assets/css/theme.scss#L13).

```
/assets
├── css
|  ├── admin.scss            => Arquivo SCSS principal Admin WP
|  ├── editor.scss           => Arquivo SCSS principal Editor WP
|  ├── theme
|  |  ├── _abstracts.scss    => Váriaveis + Functions + Mixins do Tema, Bootstrap e etc
|  |  └── _variables.scss    => Váriaveis SCSS do tema
|  ├── theme.scss            => Arquivo SCSS principal Tema
|  └── vendors
|     ├── _bootstrap.scss    => Bootstrap personalizado
|     └── _font-awesome.scss => FontAwesome personalizado
├── fonts
├── img
|  ├── admin-logo-login.png
|  ├── theme-icon.png
|  └── theme-logo.png
└── js
├── admin.js                 => Arquivo Script principal Admin WP
└── theme.js                 => Arquivo Script principal Tema
```

adammacias committed:
Remove favicon

adammacias committed:
Remove no-js class CSS

adammacias committed:
Remove script html5.js

adammacias committed:
Add templates-parts directory

adammacias committed:
Delete docs directory

adammacias committed:
Slit helpers

adammacias committed:
Move files core to inc directory

adammacias committed:
Delete variable no use

adammacias committed:
Delete odin_debug() function

adammacias committed:
Move Odin_Widget_Like_Box

adammacias committed:
Rename optimize file to cleanup

adammacias committed:
Move register menu theme-support.php to new menu.php

adammacias committed:
Delete file optimize.php, new is cleanup.php

adammacias committed:
Fix include shortcode module

adammacias committed:
Update functions.php

adammacias committed:
Delete audio.php and video.php files

adammacias committed:
Move languages directory

adammacias committed:
Add bower file

adammacias committed:
Update gitignore

adammacias committed:
Refactor Gruntfile, Package and Bower files

adammacias committed:
Refactor assets directory

adammacias committed:
Update enqueue scripts

adammacias committed:
Fix task zip

adammacias committed:
Update comments in main theme files

adammacias committed:
Update version number Odin

adammacias committed:
Update navigation skiplink in header

adammacias committed:
Update header branding

adammacias committed:
Update navigation top

adammacias committed:
Update custom header image

adammacias committed:
Fix navigation top

adammacias committed:
Update theme support

adammacias committed:
Fix theme support

adammacias committed:
Update SCSS files

adammacias committed:
Grunt

adammacias committed:
Update dependencies package.json

adammacias committed:
Change path files SCSS vendors directory

adammacias committed:
Update Gruntfile.js

adammacias committed:
Remove bower package.json

adammacias committed:
Update main.js and grunt

adammacias committed:
Fix watch task

adammacias committed:
Update style theme and grunt

adammacias committed:
Improve editor-style.scss

adammacias committed:
Update import bourbon optional and grunt

adammacias committed:
Update .jshintrc to main directory

adammacias committed:
Rename templates-parts to template-parts

adammacias committed:
Update languages local

adammacias committed:
Rename like-box.php to class-odin-widget-like-box.php

adammacias committed:
Update assets admin

adammacias committed:
Rename Odin shortcodes

adammacias committed:
Rename classes helpers

adammacias committed:
Update .gitignore

adammacias committed:
Fix package.json

adammacias committed:
Rename libraries, hooks and helpers functions

adammacias committed:
Merge pull request #398 from adammacias/v3-dev

Odin v3-dev prepare

rodrigo-brito committed:
Changes in tags and missing comma

adammacias committed:
Merge pull request #400 from rodrigo-brito/fix-tags

Changes in tags and missing comma

adammacias committed:
Merge remote-tracking branch 'odin/master' into v3-dev

adammacias committed:
Fix merge conflit

adammacias committed:
Generate files by grunt

adammacias committed:
Fix text-domain

adammacias committed:
Add CSS generated WordPress core

adammacias committed:
Add content_width theme support

adammacias committed:
Fix text-domain translate theme

adammacias committed:
Fix pattern comment

adammacias committed:
Delete Gruntfile.js

adammacias committed:
Update package.json

adammacias committed:
Add gulpfile.js

adammacias committed:
Refactor scripts

adammacias committed:
Refactor styles

adammacias committed:
Refactor images

adammacias committed:
Refactor fonts

adammacias committed:
Update bower dependencies

adammacias committed:
Update enqueue_scripts() function

allysonsouza committed:
Renomeado diretório das classes de /libraries para /classes

allysonsouza committed:
Removidas classes que adentravam o território de plugins. ver - https://make.wordpress.org/themes/handbook/review/required/explanations-and-examples/#plugin-territory

allysonsouza committed:
Implementação do TGM para requerer o Odin Toolkit

allysonsouza committed:
Comentado requisição do TGM, deixando a cargo do desenvolvedor definir se o Toolkit é recomendado

allysonsouza committed:
Removido código de testes para criação de custom post type

adammacias committed:
Add command scripts npm. To can call "npm run dev"

adammacias committed:
Update devDependencies in package.json

adammacias committed:
Add yarn.lock file

adammacias committed:
Build

adammacias committed:
Merging with Odin v2.x

adammacias committed:
Add task gulp dist

adammacias committed:
Remove task deploy gulp

adammacias committed:
Update bower dependencies

adammacias committed:
Build

adammacias committed:
Merge remote-tracking branch 'odin/v3-dev' into v3-dev

allysonsouza committed:
Merge branch 'v3-dev' of https://github.com/wpbrasil/odin into v3-dev

allysonsouza committed:
Merge branch 'v3-dev' of https://github.com/HasteDesign/odin into v3-dev

allysonsouza committed:
Renamed and moved to /inc the required plugins file

allysonsouza committed:
Fixed the forgotten */ at the end of required-plugins.php

allysonsouza committed:
Removed TGMPA directly inserted and add composer
Removed TGMPA code that was inserted directly on project and created a
composer.json requiring the project. Changed the TGMPA class require dir on
required-plugins.php.

allysonsouza committed:
Removed /vendor from .gitignore

adammacias committed:
Update comments functions

adammacias committed:
Add composer autoload #448

adammacias committed:
Clear configuration TGM Plugin #448

adammacias committed:
Add config file to register the required plugins #448

allysonsouza committed:
Add /vendor to .gitignore #448

adammacias committed:
Remove CHANGELOG.md, now is Releases/Tags tab on GitHub

adammacias committed:
Remove Grunt.js

adammacias committed:
Remove fonts glyphicons bootstrap 3

adammacias committed:
Remove sass bootstrap 3

adammacias committed:
Remove scripts bootstrap 3

adammacias committed:
Remove script main bootstrap 3

adammacias committed:
Merge Add jQuery as dependency

adammacias committed:
Remove odin-livereload

adammacias committed:
Remove enqueue script bootstrap 3

adammacias committed:
Remove enqueue script fitvids

adammacias committed:
Remove html5shiv - bye support lt IE 9

adammacias committed:
Improve odin_enqueue_scripts function

adammacias committed:
Improve odin_stylesheet_uri function

adammacias committed:
Update name style and script files

adammacias committed:
Add webfonts fontawesome 5

adammacias committed:
Add styles fontawesome 5

adammacias committed:
Add Bootstrap 4

adammacias committed:
Remove no-js tag html, not using Modernizr more

adammacias committed:
Update navigation skiplink component

adammacias committed:
Update meta tag viewport bootstrap 4

adammacias committed:
Update navigation-main component

adammacias committed:
Update header component

adammacias committed:
Add brand-logo component

adammacias committed:
Rename navigation components

adammacias committed:
Update footer component

adammacias committed:
Improve header component

adammacias committed:
Add post compoent

adammacias committed:
Add post page compoent

adammacias committed:
Remove old content, now is components/post

adammacias committed:
Update styles structure

adammacias committed:
Fix warning excerpt in header component

adammacias committed:
Add feature functions for media display

adammacias committed:
Add page-title component

adammacias committed:
Add page-description component

adammacias committed:
Add banner component

adammacias committed:
Update header component

adammacias committed:
Update navigation-main component

adammacias committed:
Update header component

adammacias committed:
Add content-404 component and update 404 template

adammacias committed:
Remove class-bootstrap-nav.php is not necessary more

adammacias committed:
Improve banner component

adammacias committed:
Add pagination posts component

adammacias committed:
Add pagination between single posts component

adammacias committed:
Add pagination within post component

adammacias committed:
Update sidebar component

adammacias committed:
Update form search component

adammacias committed:
Improve post component

adammacias committed:
Add app component

adammacias committed:
Rename content to main component

adammacias committed:
Delete class-bootstrap-nav.php file, is not necessary more

adammacias committed:
Fix sticker footer classes css

adammacias committed:
Update index template

adammacias committed:
Update page template

adammacias committed:
Update page sidebar template

adammacias committed:
Update search template

adammacias committed:
Add native wordpress styles sass

adammacias committed:
Update 404 template

adammacias committed:
Update archive template

adammacias committed:
Fix classes css in footer

adammacias committed:
Update author template

adammacias committed:
Remove category template

adammacias committed:
Remove audio template

adammacias committed:
Remove image template

adammacias committed:
Update single template

adammacias committed:
Remove tag template

adammacias committed:
Remove video template

adammacias committed:
Add header and foote parts of post

adammacias committed:
Add navigation-top component

adammacias committed:
Add audio, gallery and video components of formats post

adammacias committed:
Add post none component

adammacias committed:
Delete html5 shiv, ie 9 is dead

adammacias committed:
Refactor header and banner components

adammacias committed:
Uncomment post-formats theme support

adammacias committed:
Refactor styles

adammacias committed:
Refactor scripts

adammacias committed:
Improve banner component

adammacias committed:
Remove fitvids.js file

adammacias committed:
Refactor page component

adammacias committed:
Update post component

adammacias committed:
Remove favicon.ico

adammacias committed:
Remove search template

adammacias committed:
Improve functions sidebar component

adammacias committed:
Improve widgets sidebar component

adammacias committed:
Add widgets in footer component

adammacias committed:
Fix id in sidebar component

adammacias committed:
Refactor comments functions, templates and styles

adammacias committed:
Remove template-tags, don't using more

adammacias committed:
Refactor functions, includes and core odin

adammacias committed:
Add dist and vendor in .gitignore file

adammacias committed:
Feature gulp - update package.json

adammacias committed:
Feature gulp - add gulpfile.js

adammacias committed:
Update style.css new version

adammacias committed:
Update README.md

adammacias committed:
Improve brand-logo functions

adammacias committed:
Rename src to assets directory

adammacias committed:
Remove scroll link in banner component

adammacias committed:
Update styles banner, form-search, brand and navigation components

adammacias committed:
Remove odin_banner_component_enqueue_script() function, scroll link not exists more

adammacias committed:
Update and compress screenshot

adammacias committed:
Update brand-logo component

adammacias committed:
Fix comment navigation component

adammacias committed:
Improve functions custom-header

adammacias committed:
Remove webfonts source by font-awesome, because is better get from npm

adammacias committed:
Update gulpfile.js

adammacias committed:
Update .jshintrc file

adammacias committed:
Update path inc files

adammacias committed:
Update .editorconfig file to WordPress coding standards

adammacias committed:
Improve functions post

adammacias committed:
Add customize selective refresh widgets

adammacias committed:
Merge branch 'v3-dev' of github.com:wpbrasil/odin into v3-beta

# Conflicts:
#	.gitignore
#	404.php
#	archive.php
#	assets/css/editor-style.css
#	assets/js/main.js
#	assets/js/main.min.js
#	author.php
#	category.php
#	comments.php
#	core/assets/css/admin.css
#	core/assets/js/admin-custom-status.js
#	core/assets/js/admin.js
#	core/assets/js/editor-shortcodes.js
#	core/classes/class-shortcodes-menu.php
#	core/classes/class-shortcodes.php
#	core/classes/class-thumbnail-resizer.php
#	core/classes/widgets/class-widget-like-box.php
#	footer.php
#	functions.php
#	gulpfile.js
#	header.php
#	image.php
#	inc/admin/config.php
#	inc/classes/class-bootstrap-nav.php
#	inc/classes/class-thumbnail-resizer.php
#	inc/cleanup.php
#	inc/core/assets/css/admin.css
#	inc/core/assets/images/odin-thumb-placeholder.jpg
#	inc/core/assets/images/placeholder.png
#	inc/core/assets/js/admin-custom-status.js
#	inc/core/assets/js/admin.js
#	inc/core/assets/js/editor-shortcodes.js
#	inc/core/classes/abstracts/abstract-front-end-form.php
#	inc/core/classes/class-contact-form.php
#	inc/core/classes/class-metabox.php
#	inc/core/classes/class-options-helper.php
#	inc/core/classes/class-post-form.php
#	inc/core/classes/class-post-status.php
#	inc/core/classes/class-post-type.php
#	inc/core/classes/class-shortcodes-menu.php
#	inc/core/classes/class-shortcodes.php
#	inc/core/classes/class-taxonomy.php
#	inc/core/classes/class-term-meta.php
#	inc/core/classes/class-theme-options.php
#	inc/core/classes/class-thumbnail-resizer.php
#	inc/core/classes/class-user-meta.php
#	inc/core/classes/widgets/class-widget-like-box.php
#	inc/core/odin-shortcodes-editor-i18n.php
#	inc/css/odin-admin.css
#	inc/js/admin-custom-status.js
#	inc/js/editor-shortcodes.js
#	inc/js/odin-admin.js
#	inc/odin-shortcodes-editor-i18n.php
#	inc/shortcodes/class-odin-shortcodes-menu.php
#	inc/shortcodes/class-odin-shortcodes.php
#	inc/shortcodes/odin-shortcodes-editor-i18n.php
#	inc/template-tags.php
#	inc/theme/functions.php
#	inc/widgets/class-odin-widget-like-box.php
#	index.php
#	package.json
#	page-sidebar.php
#	page.php
#	search.php
#	searchform.php
#	sidebar.php
#	single.php
#	src/img/logo-login.svg
#	src/scss/admin-style.scss
#	src/scss/shared/_wordpress.scss
#	style.css
#	tag.php
#	template-parts/content-page.php
#	template-parts/content.php

adammacias committed:
Remove odin_pagination(), use the_posts_pagination()

adammacias committed:
Fix text domain

adammacias committed:
Update version Odin

adammacias committed:
Add task and package gulp-checktextdomain

adammacias committed:
Add task and package node-wp-i18n

adammacias committed:
Add odin.pot file in languages

adammacias committed:
Update translate pt_BR theme

adammacias committed:
Add build:fonts command

adammacias committed:
Fix class css pagination component

adammacias committed:
Update Bootstrap v4.0 to v4.1

adammacias committed:
Update FontAwesome v5.0.7 to v5.0.10

adammacias committed:
Update node-wp-i18n and gulp-sass packages

allysonsouza commented:
Testei por aqui @adammacias, tudo certo! Achei bem interessante a estruturação, a primeira vista tá bem fácil de compreender, e aparentemente produtivo para desenvolver, mas aí só pegando pra fazer um projeto pra sentir. Vamos ver o que o resto da galera acha!

clcmoliveira commented:
Vou testar em meu WAMP assim que puder. Mas parece estar muito bom só pelas imagens.

chrdesigner commented:
Eu encontrei um "erro" com integração com o plugin odin-toolkit + Odin v3. Ele solicita a pasta "core" aonde estão o css + js para criação dos campos personalizados... A minha duvida é se criamos uma pasta dentro do plugin "que eu acho muito mais viável" ou deixamos na pasta do tema :)

allysonsouza commented:
@chrdesigner o Odin Toolkit ficou meio que no limbo, precisa de uns updates, tem vários bugzinhos mesmo, acabei seguindo com um fork do projeto enquanto as coisas no Odin avançavam a passos lentos: https://github.com/HasteDesign/Haste-Toolkit

Mas com essa renovada da v3, vou tentar dar um tapa no Toolkit também

allysonsouza committed:
Commenting TGM activation
- Required and recommended plugins should start commented and be activated if needed

AdsonCicilioti committed:
fix para node v10

AdsonCicilioti committed:
update bootstraop v4.1.3

AdsonCicilioti committed:
To fix dependencies on Windows

adammacias committed:
Merge pull request #487 from AdsonCicilioti/v3-beta

Cross platform compatibility

gspmonteiro commented:
> Demorei mas testei tb..rs
> 
> Encontrei apenas um problema no menu. Pois ele aparece "desconfigurado".
> 
> ![screenshot from 2018-06-09 10-44-04](https://user-images.githubusercontent.com/6010232/41192313-29e7bf70-6bd2-11e8-9a7a-61a9aaf16c7f.png)
> 
> Quando fui na parte administrativa para alterar o menu, recebi a seguinte mensagem:
> 
> ![screenshot from 2018-06-09 10-44-25](https://user-images.githubusercontent.com/6010232/41192324-4619cfd0-6bd2-11e8-986a-0529c546ff0d.png)
> 
> Verifiquei que no código não é feito o registro do menu ([register_nav_menu](https://developer.wordpress.org/reference/functions/register_nav_menu/)). Ao fazer esse registro a mensagem desaparece, mas o menu fica da mesma forma.

Olá, sou novo por aqui,
Eu tbm percebi isso, mas é só registrar com o mesmo theme_location que se encontra no /components/navigation-main.php; Bom eu registrei o menu, colocando esse código abaixo no /inc/theme/functions.php: 
```
	       /**
		 * Register nav menus.
		 */
		register_nav_menus(
			array(
				'main-menu' => __( 'Main Menu', 'odin' )
			)
		);
```
Minha única crítica é que ficar mascarando o uso do bootstrap pode ser mais chato que simplesmente usar as classes esplicitamente, por exemplo: eu quis substituir o search_form do topbar por um navbar-nav, então eu acabo duplicando alguns estilos que poderiam estar sendo usados uma unica vez. 

Encontrei um outro "erro" no _wordpress.scss na linha 92 temos essa declaração: 
```
.hentry,
.comment,
#comments {
	word-wrap: break-word;
	overflow-wrap: break-word;
	word-break: break-all;
	hyphens: auto;
} 
```

Porém, com essa declaração quebra-se as palavras em qualquer lugar e nem coloca o hifen de quebra "-", para consertar isso é só trocar o `word-brake: keep-all`, aí ele coloca as hifens no lugar certo, mas só se justificaria se tiver o `text-align:justify` todos os textos, senão a aprência fica feia.

hansmosl committed:
Adição do hook wp_body_open() no Tema

Conforme já indicado em #493

rudwolf committed:
Merge pull request #494 from hansmosl/patch-10

Adição do hook wp_body_open() no Tema

rudwolf commented:
@allysonsouza @anyssa @adammacias 

Como tá o andamento disto?

Eu aprovei 3 PRs do @hansmosl que tinham mudanças menores com atualizações de filtros deprecados e inclusão de tags novas, nada que vá criar uma enorme diferença, mas causam conflito com esta nova versão que tá pra sair, tem previsão de release dela ou é só um projeto mesmo?
